src/main/kotlin/build/buf/gradle/ProtobufGradlePluginSupport.kt

```diff
@@ -108,7 +108,7 @@ private fun Task.allProtoDirs(): List<Path> =
     (project.srcProtoDirs() + extractProtoDirs()).filter { project.anyProtos(it) }
 
 internal fun Project.srcProtoDirs() =
-    the<SourceSetContainer>().flatMap { it.protoDirs(this) } + androidSrcProtoDirs()
+    (the<SourceSetContainer>().flatMap { it.protoDirs(this) } + androidSrcProtoDirs()).minus(extractProtoDirs())
```

Comment:
Poking around possibly mindlessly, does [this](https://github.com/google/protobuf-gradle-plugin/blob/9d2a328a0d577bf4439d3b482a953715b3a03027/src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy#L425) seem to be the cause?

It looks like the code from the PR mentioned in my last comment changed the proto sourcesets to be the output of the extract protos task, which I assume is `extracted-protos`.

Reply:
Looks like that's the problem although I haven't dug in completely. If they've changed their model of the proto source set, then this plugin should do the same. It might make sense to merge this as a stop-gap, but the discrepancy will only get worse. This plugin is bound to p-g-p's implementation details - it has to be - and they've changed a fundamental part of their design, so we have to adjust.

Reply:
Updated the comment and requested. Please take a look / let me know if there's something else that should be done (followup issue, etc).

Reply:
I think I see what they did now: where previously they manually [included the `extractProtosDirs`](https://github.com/google/protobuf-gradle-plugin/pull/637/files#diff-3405f98848db3cae29f5eda1db42d03fe57eeedc8191b5bf2dd2bbfd08537240L394) and the [`src/main/proto` (and other new source) dirs](https://github.com/google/protobuf-gradle-plugin/pull/637/files#diff-3405f98848db3cae29f5eda1db42d03fe57eeedc8191b5bf2dd2bbfd08537240L392), now they just [include the `protoSourceSet.proto` dirs](https://github.com/google/protobuf-gradle-plugin/pull/637/files#diff-3405f98848db3cae29f5eda1db42d03fe57eeedc8191b5bf2dd2bbfd08537240R383) and [add the `extract` dir](https://github.com/google/protobuf-gradle-plugin/pull/637/files#diff-3405f98848db3cae29f5eda1db42d03fe57eeedc8191b5bf2dd2bbfd08537240R411) to it [along with its normal source dirs](https://github.com/google/protobuf-gradle-plugin/pull/637/files#diff-3405f98848db3cae29f5eda1db42d03fe57eeedc8191b5bf2dd2bbfd08537240R229).

This plugin relies on the `srcDirs` where we actually want to depend only the "normal source dirs" in that last link. So this "set minus" is correct, but it'll break if they add something else to that directory as well.

I don't know the best way to unmix these two directory sets. One way is this laser-focused "set minus" since we know exactly what they're adding. Another is to try to filter by what's not in the `build` directory, since they shouldn't be putting anything outside it. I don't see a way at the moment to do better than that.

Reply:
I think minus is a fine solution that can be revised if needed in the future. I think the names of some of these extension functions should change to reflect their intent.

- `srcProtoDirs` should be `nonExtractedProtoSrcDirs`
- `androidSrcProtoDirs` should be `nonExtractedAndroidSrcProtoDirs`
- `protoDirs` should be `protoSrcDirs` with a comment indicating this now includes the `extracted-protos` dir. Probably move and adapt your excellent comment from `srcProtoDirs`?
- `extractProtoDirs` should be `allExtractedProtoDirs` (notably not `allExtractedProtoSrcDirs`) and should be 

The `minus` should be:
```
(the<SourceSetContainer>().flatMap { it.protoDirs(this) } + androidSrcProtoDirs()) - Paths.get(BUILD_EXTRACTED_PROTOS_MAIN)
```

since the extracted-include protos should not cause this issue.

I'm a bit split on if your comment belongs on the `minus` or on `protoSrcDirs` or if there should be a short comment on each.

Reply:
> Are you suggesting...

That is the right extrapolation from what I was saying, but I didn't quite get it right. You're right that the Android function should not be renamed without the minus included itself, but that's probably not the right approach.

Trying again:

- `srcProtoDirs` to `protoSourceSetDirsWithoutExtractedProtos`
- `androidSrcProtoDirs` to `androidProtoSourceSetDirs`
- `protoDirs` to `protoSourceSetDirs`
- `extractProtoDirs` should actually probably be refactored and perhaps removed: its only existing usage in `allProtoDirs` should be unneeded. `Task.allProtoDirs` should just be the union of the "un-minused" proto source set dirs (`srcProtoDirs` as it exists today).

> One final question...

The driving logic is:
- Buf needs a complete proto schema in order to run its analyses
- Therefore the plugin must unify the user's code with p-g-p's extracted code, which comes in two flavors:
  - extracted-protos, schemas for which to generate code from dependencies), and
  - extracted-include-protos, schema code that needs to be present for protoc dependency resolution and therefore also Buf compilation
- For this purpose this plugin builds an ephemeral workspace that uses symlinks to point Buf to all the relevant proto directories.
- However, we should not be analyzing (linting, formatting, etc.) code from third party dependencies. So for the purpose of Buf analysis this plugin must resolve "project original" code. It used to be sufficient to take the code in the `proto` SourceSet extensions since that didn't include the `extract` code, but now it does.

Reply:
So one thing to note:

The unminused proto-dirs **do not** include the `extracted-include-protos`  directory. It looks like the `protobuf-gradle-plugin` was updated to include all directories that have `protoc` called over them. The usage of `allProtoDirs` is to both generate symlinks in the `bufbuild` directory, as well as populate the workspace yaml to let buf know where all the protos to consider exist, which needs to include `extracted-include-protos`. So I think like `allProtoDirs` will need to be something like:

```
srcProtoDirs (as it exists today) + Paths.get(BUILD_EXTRACTED_INCLUDE_PROTOS_MAIN)
```

Does that look right? We can likely remove `extractProtoDirs` as you say, as there doesn't seem to be an occasion to need it anymore.

Reply:
Oh yes... I just realized that we should probably leave the addition of `extractProtoDirs` in as the version of `protobuf-gradle-plugin` in use may vary. Versions pre `0.9.2` of the `protobuf-gradle-plugin` will break if used with this code.

I'll push an update in a bit that I think will work with old and new versions of the `protobuf-gradle-plugin`.

Reply:
It is OK for this plugin to require a specific version of the protobuf-gradle-plugin - it's pre-1.0, so any consumers, including anyone who uses this plugin, must be OK with incompatibilities between versions.

> I think like allProtoDirs will need to be something like

Yes, I think this is correct for "what we provide to Buf to build an image." The real problem is that we're reusing that logic (the function is `internal` to facilitate that) where we shouldn't be anymore, for linting and formatting. But the naming must improve along with the changed logic. In my previous comment I started in the other direction: given better naming and slightly tweaking the building blocks that are there already, can we figure out how they should be put together? In that world `protoSourceSetDirsWithoutExtractedProtos` is what we should provide to the lint and format commands.

[Here is a patch](https://github.com/bufbuild/buf-gradle-plugin/files/11672819/p-g-p-0-9-3.patch) with changes that I think encapsulate what I'm talking about (and enable all tests to pass). It's a bit nasty and has some leftover debug logging I was doing, and far too many calls to `Path.relativize`. Feel free to clean it up and incorporate it or try another approach, or I can work it into a patch for a separate PR.

I'm sure there's more than one good way to put it all together, but I think it's essential that the concepts are conveyed clearly for the next time something goes awry.

Reply:
Alrighty! Incorporated your changes with some minor tweaks w.r.t. helper names, comments, etc.

Please take a look!